### ISSUE TITLE
Fitting the model directly with `fit_gpytorch_scipy/torch` leads to buggy behavior with input transforms.

### ISSUE THREAD

**Issue by saitcakmak** [pytorch/botorch#893]
## Issue description

Since the 0.5 release, one has to call `_set_transformed_inputs` after model training to update `model.train_inputs` with their transformed versions (think `model.train_inputs = model.transform_inputs(model.train_inputs)`). This is done at the end of `fit_gpytorch_model`, so everything works great if one uses `fit_gpytorch_model` to train their model. 

But if one instead uses `fit_gpytorch_scipy/torch` directly to fit the model (e.g., #874 - saw other examples of this as well), `_set_transformed_inputs` does not get called automatically, the user is probably not aware of this need either, so we end up with a buggy behavior. Without `_set_transformed_inputs`, we end up calculating the posterior using non-transformed training inputs with transformed test inputs, along with hyper-parameters that were trained using transformed training inputs.

## Code example

```
import torch
from botorch import fit_gpytorch_model
from botorch.models import SingleTaskGP
from botorch.models.transforms.input import Normalize
from botorch.optim.fit import fit_gpytorch_torch
from gpytorch import ExactMarginalLogLikelihood


scale = 5
t_X = torch.linspace(0, scale, 10).view(-1, 1)
t_Y = torch.randn(10, 1)
model = SingleTaskGP(t_X, t_Y, input_transform=Normalize(d=1))
mll = ExactMarginalLogLikelihood(model.likelihood, model)
fit_gpytorch_model(mll)
print("Train inputs with `fit_gpytorch_model`: ", model.train_inputs)

model = SingleTaskGP(t_X, t_Y, input_transform=Normalize(d=1))
mll = ExactMarginalLogLikelihood(model.likelihood, model)
fit_gpytorch_torch(mll)
print("Train inputs with `fit_gpytorch_torch`: ", model.train_inputs)
```

Output (we want to see them normalized here):
```
Train inputs with `fit_gpytorch_model`:  (tensor([[0.0000],
        [0.1111],
        [0.2222],
        [0.3333],
        [0.4444],
        [0.5556],
        [0.6667],
        [0.7778],
        [0.8889],
        [1.0000]]),)

Train inputs with `fit_gpytorch_torch`:  (tensor([[0.0000],
        [0.5556],
        [1.1111],
        [1.6667],
        [2.2222],
        [2.7778],
        [3.3333],
        [3.8889],
        [4.4444],
        [5.0000]]),)
```

## How to fix?

- We could move `_set_transformed_inputs` into `fit_gpytorch_scipy/torch`. Existing users of `fit_gpytorch_scipy/torch` would be very happy that they don't have to change anything :)
- We could yell at the user that they have to use `_set_transformed_inputs` if using `fit_gpytorch_scipy/torch` directly. We may have to then filter out this warning if they're used from within `fit_gpytorch_model`.
- We could somehow prevent the user from using `fit_gpytorch_scipy/torch` directly and force them to use `fit_gpytorch_model`.

Thoughts @wjmaddox, @sdaulton, @Balandat ?

**Comment by sdaulton** [comment]
I think moving _set_transformed_inputs down to `fit_gpytorch_scipy/torch` is a good solution

**Comment by Balandat** [comment]
> I think moving _set_transformed_inputs down to fit_gpytorch_scipy/torch is a good solution

Yeah of the given ones that seems to be the most reasonable one. However, it's totally reasonable to use other, non-canned ways of optimizing the model parameters (e.g. writing a custom loop), in which case this would not work properly. 

How about modifying the `eval` call in such a way that `_set_transformed_inputs` is transformed automatically whenever `eval` is called? That way things should work transparently.

**Comment by wjmaddox** [comment]
In that setting, we would then require people sub-classing Botorch Model classes to be able to use input transforms, as you can currently use Botorch acquisition functions + transforms based off of just writing a batched posterior function on your model (say a gpytorch one). 

This also requires every posterior function ensuring that there is a `self.eval()` call in the model as well (which is probably fine).

**Comment by saitcakmak** [comment]
> How about modifying the eval call in such a way that _set_transformed_inputs is transformed automatically whenever eval is called? That way things should work transparently for all cases without gotcha.

I implemented a version of this in #894. With that implementation, if `model` is in `eval` mode, the `train_inputs` will be `transformed` and they'll be at their original values whenever `model` is in `train` mode. It's very much agnostic to the optimizer used, and should work with any model that subclasses `Model`. If the `model` does not have `train_inputs` and still uses `input_transform`, it is left to the model to disable the `_set_transformed_inputs` and `_revert_to_original_inputs` as is done with `DeterministicModel`.

> In that setting, we would then require people sub-classing Botorch Model classes to be able to use input transforms,

I see your point, but I think it is reasonable to expect people to either subclass `Model` or otherwise build in sufficient support into their custom model if they want to use `input_transform`. This is a BoTorch-only feature that is (sadly) not supported on the GPyTorch end.

> This also requires every posterior function ensuring that there is a self.eval() call in the model as well (which is probably fine).

Yeah, to properly apply input transforms, you need to have `if self.training: X = self.transform_inputs(X)` in the `forward` call and `self.eval(); X = self.transform_inputs(X)` in the `posterior` call (+ need to call `_set_transformed_input` after training). This is true even without the proposed change. (Also, if you don't have `self.eval()` in the `posterior`, GPyTorch will complain that you're not training on train inputs - for `ExactGP`s.)

I guess you could still write a custom model with only a `self.input_transform(X)` in `forward` pass but this would only support a limited subset of input transforms. If you only want to use `Normalize`, I think this is still fine but in this case you can't subclass `Model` (after #894).

Overall, I find the #894 design to be the lesser evil. Happy to discuss other alternatives as well though. Ideally, we would upstream the transforms (cornellius-gp/gpytorch#1652) so that we wouldn't have any gotchas and transforms would work nicely with both ExactGP and ApproximateGP.

**Comment by Balandat** [comment]
I think the basic `Model` API should be simple enough so that it's straightforward to subclass any custom model from it, so it's fine to expect that. I think #894 is a relatively clean design that removes the need for doing custom stuff in the optimizers, so that's removing some gotchas already.

I agree that the right long-term solution would be to upstream the input transforms to gpytorch.

**Comment by saitcakmak** [comment]
Closed by #894
